src/rust/src/backend/ec.rs

```diff
@@ -0,0 +1,570 @@
+
+        let py_private_key = utils::bn_to_py_int(py, ec.private_key())?;
+
+        let ec_mod = py.import(pyo3::intern!(
+            py,
+            "cryptography.hazmat.primitives.asymmetric.ec"
+        ))?;
+
+        let public_numbers = ec_mod.call_method1(
+            pyo3::intern!(py, "EllipticCurvePublicNumbers"),
+            (py_x, py_y, self.curve.clone_ref(py)),
+        )?;
+
+        Ok(ec_mod.call_method1(
+            pyo3::intern!(py, "EllipticCurvePrivateNumbers"),
+            (py_private_key, public_numbers),
+        )?)
+    }
+
+    fn private_bytes<'p>(
+        slf: &pyo3::PyCell<Self>,
+        py: pyo3::Python<'p>,
+        encoding: &pyo3::PyAny,
+        format: &pyo3::PyAny,
+        encryption_algorithm: &pyo3::PyAny,
+    ) -> CryptographyResult<&'p pyo3::types::PyBytes> {
+        utils::pkey_private_bytes(
+            py,
+            slf,
+            &slf.borrow().pkey,
+            encoding,
+            format,
+            encryption_algorithm,
+            true,
+            false,
+        )
+    }
+}
+
+#[pyo3::prelude::pymethods]
+impl ECPublicKey {
+    #[getter]
+    fn key_size<'p>(&'p self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        self.curve.as_ref(py).getattr(pyo3::intern!(py, "key_size"))
+    }
+
+    fn verify(
+        &self,
+        py: pyo3::Python<'_>,
+        signature: &[u8],
+        data: &pyo3::types::PyBytes,
+        signature_algorithm: &pyo3::PyAny,
+    ) -> CryptographyResult<()> {
+        let ecdsa_class: &pyo3::types::PyType = py
+            .import(pyo3::intern!(
+                py,
+                "cryptography.hazmat.primitives.asymmetric.ec"
+            ))?
+            .getattr(pyo3::intern!(py, "ECDSA"))?
+            .extract()?;
+
+        if !signature_algorithm.is_instance(ecdsa_class)? {
+            return Err(CryptographyError::from(
+                exceptions::UnsupportedAlgorithm::new_err((
+                    "Unsupported elliptic curve signature algorithm",
+                    exceptions::Reasons::UNSUPPORTED_PUBLIC_KEY_ALGORITHM,
+                )),
+            ));
+        }
+
+        let (data, _): (&[u8], &pyo3::PyAny) = py
+            .import(pyo3::intern!(
+                py,
+                "cryptography.hazmat.backends.openssl.utils"
+            ))?
+            .call_method1(
+                pyo3::intern!(py, "_calculate_digest_and_algorithm"),
+                (
+                    data,
+                    signature_algorithm.getattr(pyo3::intern!(py, "algorithm"))?,
+                ),
+            )?
+            .extract()?;
+
+        let mut verifier = openssl::pkey_ctx::PkeyCtx::new(&self.pkey)?;
+        verifier.verify_init()?;
+        let valid = verifier.verify(data, signature).unwrap_or(false);
+        // TODO: Empty the error stack. BoringSSL leaves one in the event of
+        // signature validation failure. Upstream to rust-openssl?
+        openssl::error::ErrorStack::get();
```

Comment:
Should we annotate this for just boring?

Reply:
Done, but also, what do you think about upstreaming?

Reply:
I think we should upstream. Without doing so every caller needs to know that they need this conditional or else their error stack gets polluted and who knows what their error handling does.